pkgs/tools/filesystems/bees/bees-service-wrapper

```diff
@@ -0,0 +1,206 @@
+    # mode, so config file support is not needed).
+    [[ $fs_spec = */* ]] || readConfigFileIfExists "$bees_config_dir/$fs_spec.conf"
+  fi
+
+  [[ $bees_uuid ]] || {
+    # if bees_uuid is not in our .conf file, look it up with findmnt
+    read -r bees_uuid fstype < <(findmnt -n -o uuid,fstype "$fs_spec") && [[ $fstype ]] || exit
+    [[ $fstype = btrfs ]] || die "Device type is $fstype, not btrfs"
+  }
+
+  [[ $bees_uuid = */* ]] || readConfigFileIfExists "$bees_config_dir/UUID=$bees_uuid.conf"
+
+  # Honor any values read from config files above; otherwise, set defaults.
+  bees_work_dir="${bees_work_dir:-.beeshome}"
+  bees_run_dir="${bees_run_dir:-/run/bees}"
+  bees_mnt_dir="${bees_mnt_dir:-$bees_run_dir/mnt/$bees_uuid}"
+  bees_home="${bees_home:-$bees_mnt_dir/$bees_work_dir}"
+  bees_status="${bees_status:-${bees_run_dir}/$bees_uuid.status}"
+  bees_verbosity="${bees_verbosity:-6}"
+  bees_db_size_mb="${bees_db_size_mb:-1024}"
+  bees_blockdev=${bees_blockdev:-"/dev/disk/by-uuid/$bees_uuid"}
+  bees_db_size="$(( bees_db_size_mb * 1024 * 1024 ))"
+
+  [[ -b $bees_blockdev ]] || die "Block device $bees_blockdev missing"
+  (( bees_db_size_mb % 16 == 0 )) || die "DB size must be divisible by 16MB"
+}
+
+do_run() {
+  local db old_db_size
+
+  _setup "$@"
+  mkdir -p -- "$bees_mnt_dir" || exit
+
+  # subvol id 5 is reserved for the root subvolume of a btrfs filesystem.
+  mountpoint -q "$bees_mnt_dir" || mount -osubvolid=5 -- "$bees_blockdev" "$bees_mnt_dir" || exit
+  if [[ -d $bees_home ]]; then
+    btrfs subvolume show "$bees_home" >/dev/null 2>&1 || die "$bees_home exists but is not a subvolume"
+  else
+    btrfs subvolume create "$bees_home" || exit
+  fi
+  db=$bees_home/beeshash.dat
+  touch -- "$db"
+
+  old_db_size=$(stat -c %s -- "$db")
+  new_db_size=$bees_db_size
+
+  if (( old_db_size != new_db_size )); then
+    rm -f -- "$bees_home"/beescrawl."$bees_uuid".dat
+    truncate -s "$new_db_size" -- "$db" || exit
+  fi
+  chmod 700 -- "$bees_home"
+
+  # BEESSTATUS and BEESHOME are the only variables handled by the legacy
+  # wrapper for which getenv() is called in C code.
+  BEESSTATUS=$bees_status BEESHOME=$bees_home exec "${beesd_bin:-/lib/bees/bees}" \
+    --verbose "$bees_verbosity" \
+    "$bees_mnt_dir" "$@" || exit
+}
+
+do_cleanup() {
+  _setup "$@"
+  mountpoint -q "$bees_mnt_dir" && umount -l -- "$bees_mnt_dir" || exit
+}
+
+(( $# >= 2 )) || usage
+declare -f "do_$1" >/dev/null 2>&1 || usage
+mode=$1; shift # must be a do_* function; currently "run" or "cleanup"
+
+declare -a args=( "$1" ); shift  # pass first argument (config-name|fs-spec) through literally
+
+# parse other arguments as key=value pairs, or pass them through literally if they do not match that form.
+# similarly, any option after "--" will be passed through literally.
+declare -A options=( )
+for piece; do
+  if [[ $piece = *=* ]]; then
+    options[${piece%%=*}]=${piece#*=}
```

Comment:
I wonder if it is better just to set these as globals. Then the option renaming loop could be refactored into a function and only written once…

Reply:
I'm actually more tempted to go the other way 'round and stop using globals. That would mean no longer `source`ing the configuration files, so I'll need to look back at the historical examples and documentation from the bees project; if they only documented `key=value` files, directly loading those keys into the `options` associative array stops them from modifying environment variables with side effects.

Reply:
Oh well, you could also use a `( )` subshell with its own private globals, then send a key-value config out of the subshell for parsing.

Reply:
A bit of context: I work in security-sensitive space where we want to prevent administrators who are authorized to make a limited set of changes (or even who can pull disks and rewrite any contents on a filesystem not flagged read-only and backed by a block device with dm_verity signatures) from being able to perform arbitrary command injection. The hackery I used earlier felt wrong when I wrote it, and it still feels wrong now -- if I don't want services I use in that context to have undocumented arbitrary-code-execution opportunities in their configuration, I shouldn't be responsible for them elsewhere.

Reply:
Maybe upstream did document the ACE…  Hopefully not.

I am ready to review a script to source the config inside `nsjail`, although it would indeed be nicer not to need such things.

Reply:
Unfortunately, it looks like historical `beesd.conf.example` files did indeed depend on being able to use shell arithmetic inside the configuration.

It's not as strong as `nsjail`, but I think we can get away with `bash -r` here (with a PATH that doesn't allow any external executables to be invoked).